### ISSUE TITLE
Redirect behaviour: when page URLs change, the previous URL should be redirected to new URL

### ISSUE THREAD

**Issue by lboaretto** [wagtail/wagtail#2618]
Automatic redirects after moving/renaming pages, tracking changes like: https://github.com/saulshanabrook/django-url-tracker would be awesome! 

It was previously discussed in #2527

**Comment by mattlinares** [comment]
Not sure if this is helpful, but adding a hook block like this does the job. There's probably a better way to integrate this into core though.

```# Create redirect when editing slugs
@hooks.register('before_edit_page')
def create_redirect_on_slug_change(request, page):
    try:
        if page.slug != request.POST['slug']:
            Redirect.objects.create(
                    old_path=page.url[:-1],
                    site=page.get_site(),
                    redirect_page=page
                )
    except:
        pass```

**Comment by ababic** [comment]
Have been looking at implementing something like this in a client project, and decided to try and utilise the `pre_save` signal for `Page` to detect:
1. When a page's slug changes
2. When a page is moved to a different section

In each case, I want to create a redirect for `instance`, and also one for each of it's descendants (because their URLs will be affected too). My signal handler looks something like this (deliberately unoptimised for now):

```
from django.db.models.signals import pre_save
from django.db.utils import IntegrityError
from wagtail.contrib.redirects.models import Redirect
from wagtail.core.models import Page

def create_redirects_on_page_move_or_slug_change(sender, instance, **kwargs):
    old_page = Page.objects.get(id=instance.pk).specific
    old_page_site = old_page.get_site()
    if(
        instance.slug != old_page.slug or
        instance.url_path != old_page.url_path
    ):
        try:
            Redirect.objects.create(
                site=old_page_site,
                old_path=old_page.get_url(current_site=old_page_site),
                redirect_page=instance,
            )
        except IntegrityError:
            pass
        for descendant in instance.get_descendants().specific():
            descendant_site=descendant.get_site()
            try:
                Redirect.objects.create(
                    site=descendant_site,
                    old_path=descendant.get_url(current_site=descendant_site),
                    redirect_page=descendant,
                )
            except IntegrityError:
                pass

pre_save.connect(
    receiver=create_redirects_on_page_move_or_slug_change,
    sender=Page
)
```

This seems to work okay for the redirect created for `instance`. However, the redirects created for the descendants seem take on the descendant's new path value for `old_path`, which doesn't seem right. It would seem that the `url_path` values of the descendant pages are updated before this signal is received, making it difficult to retrieve the descendant pages URLs before the change.

**Comment by marteinn** [comment]
There is a third party package that does this: https://github.com/themotleyfool/wagtail-automatic-redirects

**Comment by yhoiseth** [comment]
I discovered this issue after following the documented guide to [have redirects created automatically when changing page slug](https://docs.wagtail.io/en/stable/reference/pages/model_recipes.html#have-redirects-created-automatically-when-changing-page-slug). It worked well in staging and I _almost_ changed the slug in production. But then I realized that I also needed to test child pages. And it turns out that the script does not handle child pages.

[wagtail-automatic-redirects](https://github.com/themotleyfool/wagtail-automatic-redirects) handles them flawlessly, however.

I think the documentation should be updated to present wagtail-automatic-redirects as an alternative. I can make a go at it if desirable.
